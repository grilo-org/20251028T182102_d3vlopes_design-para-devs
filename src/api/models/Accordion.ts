import { Accordion } from '../generated/graphql';

export type AccordionAPI = Omit<Accordion, 'stage'>;
