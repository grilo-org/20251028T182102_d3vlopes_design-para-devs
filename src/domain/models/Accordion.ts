export interface AccordionModel {
	id: string;
	title: string;
	content: string;
}
