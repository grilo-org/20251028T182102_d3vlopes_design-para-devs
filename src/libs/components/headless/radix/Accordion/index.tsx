import {
	Content,
	Header,
	Item,
	Root,
	Trigger,
} from '@radix-ui/react-accordion';

import {
	AccordionContentWrapperProps,
	AccordionHeaderProps,
	AccordionItemWrapperProps,
	AccordionRootProps,
	AccordionTriggerWrapperProps,
} from './types';

export const AccordionRootWrapper = ({
	children,
	...props
}: AccordionRootProps): JSX.Element => {
	return <Root {...props}>{children}</Root>;
};

export const AccordionContentWrapper = ({
	children,
	ref,
	...props
}: AccordionContentWrapperProps): JSX.Element => {
	return (
		<Content ref={ref} {...props}>
			{children}
		</Content>
	);
};

export const AccordionItemWrapper = ({
	ref,
	...props
}: AccordionItemWrapperProps): JSX.Element => {
	return <Item ref={ref} {...props} />;
};

export const AccordionHeaderWrapper = ({
	children,
	...props
}: AccordionHeaderProps): JSX.Element => {
	return <Header {...props}>{children}</Header>;
};

export const AccordionTriggerWrapper = ({
	ref,
	children,
	...props
}: AccordionTriggerWrapperProps): JSX.Element => {
	return (
		<Trigger ref={ref} {...props}>
			{children}
		</Trigger>
	);
};
