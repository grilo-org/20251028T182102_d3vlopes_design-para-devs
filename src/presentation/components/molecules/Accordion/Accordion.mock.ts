import { generateID } from '@/utils/generateID';

import { AccordionProps } from '.';

export const accordionMock: AccordionProps = {
	id: generateID(),
	title: 'Accordion Title',
	content: 'Loren ipsum dolor sit amet, consectetur adipiscing elit',
};
