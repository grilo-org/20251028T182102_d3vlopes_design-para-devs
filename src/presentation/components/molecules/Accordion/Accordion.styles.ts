import { styleWrapper } from '@/libs/styles/tailwind-variants';

const styles = styleWrapper({
	slots: {
		itemStyles:
			'border-b border-neutral-500 text-neutral-50 font-highlight text-[1.5625rem] md:px-3 md:text-9xl',
		triggerWrapperStyles: 'flex',
		triggerStyles:
			'text-left flex flex-1 items-start justify-between py-4 font-bold text transition-all hover:underline [&[data-state=open]>svg]:rotate-180 md:items-center',
		iconStyles: 'h-8 w-8 shrink-0 transition-transform duration-200',
		contentWrapperStyles:
			'overflow-hidden text-base font-body text-neutral-50 transition-all data-[state=closed]:animate-accordion-up data-[state=open]:animate-accordion-down md:text-2xl md:leading-[170%]',
		contentStyles: 'pb-4 pt-0',
	},
});

const makeStyles = styles();

export const classNames = {
	itemStyles: makeStyles.itemStyles(),
	triggerWrapperStyles: makeStyles.triggerWrapperStyles(),
	triggerStyles: makeStyles.triggerStyles(),
	iconStyles: makeStyles.iconStyles(),
	contentWrapperStyles: makeStyles.contentWrapperStyles(),
	contentStyles: makeStyles.contentStyles(),
};
