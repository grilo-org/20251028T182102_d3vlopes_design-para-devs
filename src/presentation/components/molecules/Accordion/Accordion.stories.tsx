import type { Meta, StoryObj } from '@storybook/react';

import { accordionMock } from './Accordion.mock';

import { Accordion } from '.';

const meta: Meta<typeof Accordion> = {
	title: 'Molecules/Accordion',
	component: Accordion,
	args: accordionMock,
	parameters: {
		design: {
			type: 'figma',
			url: 'https://www.figma.com/file/p1r0o56xHSpG3YHnrNlO3P/LP-Design-para-Devs?type=design&node-id=143-280&mode=design&t=QuxfNOq0DxuZRe8u-0',
		},
	},
};

export default meta;
type Story = StoryObj<typeof Accordion>;

const template: Story = {
	render: (args) => (
		<div className="max-w-[76.75rem] m-auto">
			<Accordion {...args} />
		</div>
	),
};

export const Default: Story = {
	...template,
};
