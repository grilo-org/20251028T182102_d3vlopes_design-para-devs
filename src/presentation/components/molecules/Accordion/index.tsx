import { AccordionModel } from '@/domain/models/Accordion';

import { AccordionComposition } from './Accordion.composition';

export type AccordionProps = AccordionModel;

export const Accordion = ({ id, title, content }: AccordionProps) => {
	return (
		<AccordionComposition.Root
			type="single"
			collapsible
			className="px-4 md:pl-[4rem]"
		>
			<AccordionComposition.Item value={id}>
				<AccordionComposition.Title>{title}</AccordionComposition.Title>

				<AccordionComposition.Content data-testid="accordion-content">
					{content}
				</AccordionComposition.Content>
			</AccordionComposition.Item>
		</AccordionComposition.Root>
	);
};
