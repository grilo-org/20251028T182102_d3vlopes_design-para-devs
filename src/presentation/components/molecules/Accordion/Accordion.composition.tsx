import { AccordionRootWrapper } from '@/libs/components/headless/radix/Accordion';

import { AccordionContentPartial } from './partials/AccordionContentPartial';
import { AccordionItemPartial } from './partials/AccordionItemPartial';
import { AccordionTriggerPartial } from './partials/AccordionTriggerPartial';

export const AccordionComposition = {
	Root: AccordionRootWrapper,
	Item: AccordionItemPartial,
	Title: AccordionTriggerPartial,
	Content: AccordionContentPartial,
};
