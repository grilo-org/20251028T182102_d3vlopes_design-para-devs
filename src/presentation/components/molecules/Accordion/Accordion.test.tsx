import { fireEvent, render, screen } from '@testing-library/react';

import { accordionMock as mock } from './Accordion.mock';

import { Accordion } from '.';

const { getByRole, getByTestId } = screen;

describe('<Accordion />', () => {
	beforeEach(() => render(<Accordion {...mock} />));

	it('should render heading', () => {
		const heading = getByRole('heading', { name: mock.title });

		expect(heading).toBeInTheDocument();
	});

	it('should render content', () => {
		const content = getByTestId('accordion-content');

		expect(content).toBeInTheDocument();
	});

	it('should render open/close button', () => {
		const button = getByRole('button');

		expect(button);
	});

	it('should render icon', () => {
		const icon = getByTestId('accordion-icon');

		expect(icon).toBeInTheDocument();
	});

	it('should render close by default', () => {
		const content = getByTestId('accordion-content');

		expect(content).toHaveAttribute('data-state', 'closed');
	});

	it('should change close/open state on click button', async () => {
		const button = getByRole('button');
		const content = getByTestId('accordion-content');

		fireEvent.click(button);

		expect(content).toHaveAttribute('data-state', 'open');

		fireEvent.click(button);

		expect(content).toHaveAttribute('data-state', 'closed');
	});

	it('should match snapshot', () => {
		const { container } = render(<Accordion {...mock} />);

		expect(container).toMatchSnapshot();
	});
});
