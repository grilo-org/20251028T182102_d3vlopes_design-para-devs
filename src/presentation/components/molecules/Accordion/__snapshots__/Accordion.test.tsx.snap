// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Accordion /> should match snapshot 1`] = `
<div>
  <div
    class="px-4 md:pl-[4rem]"
    data-orientation="vertical"
  >
    <div
      class="border-b border-neutral-500 text-neutral-50 font-highlight text-[1.5625rem] md:px-3 md:text-9xl"
      data-orientation="vertical"
      data-state="closed"
    >
      <h3
        class="flex"
        data-orientation="vertical"
        data-state="closed"
      >
        <button
          aria-controls="radix-:rf:"
          aria-expanded="false"
          class="text-left flex flex-1 items-start justify-between py-4 font-bold text transition-all hover:underline [&[data-state=open]>svg]:rotate-180 md:items-center"
          data-orientation="vertical"
          data-radix-collection-item=""
          data-state="closed"
          id="radix-:re:"
          type="button"
        >
          Accordion Title
          <svg
            class="lucide lucide-chevron-down h-8 w-8 shrink-0 transition-transform duration-200"
            data-testid="accordion-icon"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m6 9 6 6 6-6"
            />
          </svg>
        </button>
      </h3>
      <div
        aria-labelledby="radix-:re:"
        class="overflow-hidden text-base font-body text-neutral-50 transition-all data-[state=closed]:animate-accordion-up data-[state=open]:animate-accordion-down md:text-2xl md:leading-[170%]"
        data-orientation="vertical"
        data-state="closed"
        data-testid="accordion-content"
        hidden=""
        id="radix-:rf:"
        role="region"
        style="--radix-accordion-content-height: var(--radix-collapsible-content-height); --radix-accordion-content-width: var(--radix-collapsible-content-width);"
      />
    </div>
  </div>
</div>
`;
